### PR TITLE
ddl, parser, planner: admin Pause/Resume on DDL jobs

### LINKED ISSUE #18015: Pause add an index(write reorganization) and set a time window to resume
### Description

In some scenarios, latencies of the SQL statements are very essential and should be stable all the time. However, the DDL `CREATE TABLE ... ADD INDEX` may affect the latencies of other statements: https://docs.pingcap.com/tidb/stable/online-workloads-and-add-index-operations.

For now, TiDB provided `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` system variables. By decreasing them, the resources occipued `ADD INDEX` operation can be lower, but they cannot be none.

It can be better if such features are provided:

- The ability to pause a running DDL statement like `ADD INDEX`.
- The ability to set a time-based trigger for customized SQL statements.

By the features above users are able to pause a running `ADD INDEX` manually, or schedule the `ADD INDEX` operations to make sure they will be ran at a suitable time.

### Category

Feature, Stability

### Value

Reduce latency jitter in peak periods of core application scenarios.

### PR DESCRIPTION
**Pull request by dhysum**
Add Pause/Resume admin command in parser and planner

### What problem does this PR solve?

Issue Number: close #18015 #40041

Related issue number: #39751 #43061 

Problem Summary:

### What is changed and how it works?

1. Add `admin pause ddl jobs 3,5;`
a. the job should be in Running or Queueing state
b. change the state to be Pausing on command issued, in another PR
c. background work will pause the job once scheduled, in another PR
 
2. Add `admin resume ddl jobs 3,5;`
a. the job should be paused
b. change the state to be Queueing on command issued, in another PR
c. background worker will continue the job, no need to change

### Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [x] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
- Add two commands `admin pause ddl jobs 3,5;` and `admin resume ddl jobs 3,5;`
- Add two state of the job, Pausing and Paused
- The job in Running / Queueing could be transferred to be Pausing by `admin pause ddl jobs 3,5;`, then be Paused by background worker
- Only Paused job can be resumed to be Queueing by `admin pause ddl jobs 3,5;
- New error code added: ErrCannotPauseDDLJob  = 9014, ErrCannotResumeDDLJob = 9015, ErrPausedDDLJob       = 9016
```

### CODE REVIEW

File: parser/model/ddl_test.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func TestJobSize(t *testing.T) {
 - SubJob.ToProxyJob()
 `
 	job := model.Job{}
-	require.Equal(t, 320, int(unsafe.Sizeof(job)), msg)
+	require.Equal(t, 328, int(unsafe.Sizeof(job)), msg)
```
Comment (Benjamin2037) [inline parser/model/ddl_test.go:53]: What are 320 or 328 means?

Comment (dhysum) [reply parser/model/ddl_test.go:53]: A new `int` member AdminCommandOperator has been added to struct `Job`, which increase the size of it to be with more 8 bytes.

The member AdminCommandOperator indicates where the command (cancel/pause/resume) on DDL jobs is from, end user or system. Refer to `parser.model.ddl.Job` .

File: parser/parser.y [hunk 2]
```diff
@@ -10705,6 +10705,20 @@ AdminStmt:
 			JobIDs: $5.([]int64),
 		}
 	}
+|	"ADMIN" "PAUSE" "DDL" "JOBS" NumList
```
Comment (zimulala) [inline parser/parser.y:10708]: Please add tests like https://github.com/pingcap/tidb/blob/master/parser/parser_test.go#L827.

**Comment by Benjamin2037** [comment]
The better way is add some ut and simple Ft in the pr. Normally, do not  check manual test option.

File: parser/parser.y (hunk 2 above)
Comment (dhysum) [reply parser/parser.y:10717]: Sure, on it.

Comment (dhysum) [reply parser/parser.y:10717]: done

File: parser/parser_test.go [hunk 3]
```diff
@@ -826,6 +826,10 @@ func TestDMLStmt(t *testing.T) {
 		{"admin checksum table t1, t2;", true, "ADMIN CHECKSUM TABLE `t1`, `t2`"},
 		{"admin cancel ddl jobs 1", true, "ADMIN CANCEL DDL JOBS 1"},
 		{"admin cancel ddl jobs 1, 2", true, "ADMIN CANCEL DDL JOBS 1, 2"},
```
Comment (Benjamin2037) [inline parser/parser_test.go:828]: Can we add some more test cases about exception

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: parser/parser_test.go (hunk 3 above)
Comment (dhysum) [reply parser/parser_test.go:828]: A few negative test cases added.

File: DEPS.bzl [hunk 4]
```diff
@@ -6076,8 +6076,8 @@ def go_deps():
         name = "org_modernc_golex",
         build_file_proto_mode = "disable_global",
         importpath = "modernc.org/golex",
-        sum = "h1:EYKY1a3wStt0RzHaH8mdSRNg78Ub0OHxYfCRWw35YtM=",
-        version = "v1.0.1",
```
Comment (hawkingrei) [inline DEPS.bzl:6080]: I think you should not change the DEPS.bzl because you don't update the gomod.

File: parser/model/ddl.go [hunk 5]
```diff
@@ -822,75 +939,14 @@ func (job *Job) IsRollbackable() bool {
...
-	case JobStateCancelling:
-		return "cancelling"
-	case JobStateSynced:
-		return "synced"
-	case JobStateQueueing:
-		return "queueing"
-	default:
-		return "none"
-	}
+// IsPausable checks whether we can pause the job
```
Comment (zimulala) [inline parser/model/ddl.go:942]: Ditto.

File: parser/model/ddl.go [hunk 6]
```diff
@@ -385,30 +495,27 @@ type Job struct {
...
 	RealStartTS uint64 `json:"real_start_ts"`
-	// StartTS uses timestamp allocated by TSO.
-	// Now it's the TS when we put the job to TiKV queue.
-	StartTS uint64 `json:"start_ts"`
-	// DependencyID is the job's ID that the current job depends on.
-	DependencyID int64 `json:"dependency_id"`
-	// Query string of the ddl job.
-	Query      string       `json:"query"`
+
+	// The TS when the job is pushed into TiKV queue, allocated by TSO
```
Comment (zimulala) [inline parser/model/ddl.go:514]: Ditto.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**